### PR TITLE
Only connect to Elasticsearch instances with the same version or newer

### PR DESCRIPTION
**Pull request by kvch**
## What does this PR do?

This PR adds a new callback to all Elasticsearch connections to check if the version is at least the same as the Beat. The check can be disabled by setting the new option `output.elasticsearch.allow_older_versions` to true.

## Why is it important?

By restricting the Elasticsearch versions we can provide optimal experience. 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

Closes https://github.com/elastic/beats/issues/29264

### CODE REVIEW

File: libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl [hunk 1]
```diff
@@ -77,6 +77,10 @@ output.elasticsearch:
   # Configure HTTP request timeout before failing a request to Elasticsearch.
   #timeout: 90
 
+  # {{.BeatName}} expects Elasticseach to be the same version or newer than the Beat.
```
Comment (cmacknz) [inline libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl:80]:
```suggestion
  # {{.BeatName}} expects Elasticsearch to be the same version or newer than the Beat.
```

File: libbeat/outputs/elasticsearch/client.go [hunk 2]
```diff
@@ -38,7 +38,11 @@ import (
 	"github.com/elastic/beats/v7/libbeat/testing"
 )
 
-var errPayloadTooLarge = errors.New("the bulk payload is too large for the server. Consider to adjust `http.max_content_length` parameter in Elasticsearch or `bulk_max_size` in the beat. The batch has been dropped")
+var (
+	errPayloadTooLarge = errors.New("the bulk payload is too large for the server. Consider to adjust `http.max_content_length` parameter in Elasticsearch or `bulk_max_size` in the beat. The batch has been dropped")
+
+	ErrTooOld = errors.New("Elasticsearch is too old. If you would like to connect to older instances set output.elasticsearch.allow_older_versions to true.")
```
Comment (cmacknz) [inline libbeat/outputs/elasticsearch/client.go:44]: Are these errors user facing? If they are, could we be more specific? Something like `Elasticsearch version must be greater than or equal to the beat version. If you ...."`. 

We could also wrap this and format the version into the error this so the user action is explicit like `Upgrade Elasticsearch to vX.X.X or newer: Elasticsearch is too old. If you...` or equivalent.

File: libbeat/tests/system/config/mockbeat.yml.j2 [hunk 3]
```diff
@@ -42,6 +42,8 @@ output:
     {% for k, v in elasticsearch.items() -%}
     {{ k }}: {{ v }}
     {% endfor -%}
+    # older versions have to be alloed because mockbeat is on v9.9.9
```
Comment (cmacknz) [inline libbeat/tests/system/config/mockbeat.yml.j2:45]:
```suggestion
    # older versions have to be allowed because mockbeat is on v9.9.9
```

**Comment by ruflin** [comment]
@kvch Is my understanding correct that we can target this against 8.1 as 8.0 does not talk to previous major versions anyways?

File: libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl (hunk 1 above)
Comment (cmacknz) [reply libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl]: Thanks, the templated documentation files still look like they have the typo in them. Let's not forget to update those :)

**Comment by kvch** [comment]
> @kvch Is my understanding correct that we can target this against 8.1 as 8.0 does not talk to previous major versions anyways?

We target this against 8.1. In 8.0, if someone tries to connect to an older version, we just emit a warning: https://github.com/elastic/beats/pull/29723

**Review by cmacknz — APPROVED** [review]

**Comment by ruflin** [comment]
@nimarezainia Can we make sure this is also clearly shown in our compatibility matrix? 

@kvch Based on your comment, someone could run Beats 8.0 vs ES 7.13 and the user would get a warning. Do we agree that we don't support this?

**Comment by kvch** [comment]
> @kvch Based on your comment, someone could run Beats 8.0 vs ES 7.13 and the user would get a warning. Do we agree that we don't support this?

Yes and yes.

**Comment by mgevans-5** [comment]
Please confirm this does or does not impact Logstash as the PR only mentions connections to Elasticsearch

**Comment by cmacknz** [comment]
> Please confirm this does or does not impact Logstash as the PR only mentions connections to Elasticsearch

The change only affects Elasticsearch. The only code change here is to the beats Elasticsearch client.

**Pull request merged**